// typecomposer/dist/esm/core/router/RouteView.js
import { Component, Router } from "../..";
export class RouteView extends Component {
    _view = undefined;
    _url = "";
    routePage = undefined;
    constructor(optional) {
        super(optional);
    }
    get url() {
        return this._url;
    }
    connectedCallback() {
        this.updateView();
    }
    updateView() {
        const routePage = Router.controller.getRouteViewFree(this);
        if (routePage == undefined) {
            this.view = undefined;
            return;
        }
        if (routePage == this.routePage)
            return;
        routePage.build = routePage.build || new routePage.component();
        this.routePage = routePage;
        this.view = routePage.build;
    }
    disconnectedCallback() {
        if (this.routePage)
            this.routePage.routeView = undefined;
        this.routePage = undefined;
    }
    get view() {
        return this._view;
    }
    set view(value) {
        if (this._view == value)
            return;
        if (this._view)
            this._view.remove();
        this._view = value;
        if (value)
            this?.append(value);
    }
}
// @ts-ignore
customElements.define("route-view", RouteView);
