// typecomposer/dist/cjs/core/element/action/AlertPane.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlertPane = void 0;
const __1 = require("../..");
class AlertPane extends __1.Component {
    static _div = new __1.DivElement({ id: "alert-system" });
    constructor(options) {
        super({
            className: "alert",
            display: "flex",
            flexDirection: "column",
            ...options,
        });
        const divBtn = this.appendChild(new __1.DivElement({ className: "close-btn" }));
        this.append(new __1.SpanElement({ text: options?.title || "", className: "fas fa-exclamation-circle " + options?.type }));
        this.append(new __1.SpanElement({ text: options?.message || "", className: "msg " + options?.type }));
        const btn = divBtn.appendChild(new __1.SpanElement({ className: "fas fa-times", text: "X" }));
        btn.onclick = () => this.remove();
        const divAlert = document.querySelector("#alert-system");
        if (!divAlert)
            document.body.appendChild(AlertPane._div);
        const timeout = options?.timeout || 5000;
        if (timeout > 0) {
            setTimeout(() => {
                this.remove();
            }, timeout);
        }
    }
    static success(title, message, options) {
        return AlertPane._div.appendChild(new AlertPane({ ...options, type: "success", message, title }));
    }
    static info(title, message, options) {
        return AlertPane._div.appendChild(new AlertPane({ ...options, type: "info", message, title }));
    }
    static warning(title, message, options) {
        return AlertPane._div.appendChild(new AlertPane({ ...options, type: "warning", message, title }));
    }
    static danger(title, message, options) {
        return AlertPane._div.appendChild(new AlertPane({ ...options, type: "danger", message, title }));
    }
}
exports.AlertPane = AlertPane;
customElements.define("alert-pane", AlertPane);
