// typecomposer/dist/cjs/core/components/containment/DialogPane.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DialogPane = void 0;
const __1 = require("../..");
class DialogPane extends __1.Component {
    _content = new __1.CardElement({
        className: "content",
        width: "500px",
        height: "300px",
    });
    constructor(optional) {
        super(optional);
        if (optional?.display == undefined)
            this.style.display = "none";
        this.append(this._content);
        this.addEventListener("click", (e) => {
            this.dismiss();
            e.stopPropagation();
        });
        this._content.addEventListener("click", (e) => {
            e.stopPropagation();
        });
    }
    onInit() {
        console.log("DialogElement");
    }
    get content() {
        return this._content;
    }
    show() {
        this.style.display = "block";
    }
    dismiss() {
        this.style.display = "none";
    }
}
exports.DialogPane = DialogPane;
customElements.define("dialog-pane", DialogPane);
