// typecomposer/dist/cjs/core/components/navigation/MenuBarElement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuBarElement = void 0;
const __1 = require("../..");
class MenuBarElement extends __1.Component {
    _nav;
    _ul = new __1.UListElement();
    checkbox = new __1.InputElement({
        type: "checkbox",
        id: "check",
    });
    checkboxLabel = new __1.LabelElement({
        for: "check",
        className: "checkbtn",
        text: '<i class="fas fa-bars"></i>',
    });
    _logo;
    constructor(optional) {
        super();
        if (optional?.logo) {
            this.logo = optional.logo;
            this.append(this.logo);
            delete optional.logo;
        }
        __1.Component.applyData(optional, this);
        this._nav = new __1.NavElement(optional);
        this.nav.append(this.checkbox, this.checkboxLabel, this._ul);
        this.append(this.nav);
        this.checkbox.addEventListener("change", () => {
            console.log(this.checkbox.checked);
            this._ul.style.top = this.clientHeight + "px";
        });
    }
    get nav() {
        return this._nav;
    }
    // get label(): LabelElement | undefined {
    //     return this._label;
    // }
    get logo() {
        return this._logo;
    }
    set logo(value) {
        if (this._logo)
            this._logo.remove();
        if (typeof value === "string")
            // @ts-ignore
            this._logo = new __1.LabelElement({ text: value });
        else
            this._logo = value;
        this.logo.classList.add("logo");
        this.append(this.logo);
    }
    addMenu(menu, isLogo = false, optional) {
        const li = new __1.ListItemElement();
        if (this._ul.children.length === 0)
            li.classList.add("logo");
        if (isLogo && !(typeof menu === "string")) {
            li.appendChild(menu);
        }
        else {
            const a = new __1.AnchorElement(optional);
            if (typeof menu === "string") {
                a.text = menu;
            }
            else {
                a.append(menu);
            }
            li.appendChild(a);
        }
        this._ul.appendChild(li);
    }
}
exports.MenuBarElement = MenuBarElement;
// @ts-ignore
customElements.define("menu-bar-element", MenuBarElement);
