// typecomposer/dist/esm/core/element/action/TabItem.js
import { Component, H4Element, IconElement } from "..";
export class TabItem extends Component {
    title;
    __closeable = false;
    titleElement;
    displayContent;
    _content;
    closeIcon = null;
    constructor(title, content, icon, closeable = true) {
        super();
        this.title = title;
        if (icon != null) {
            icon.style.order = "0";
            this.appendChild(icon);
        }
        this._content = content;
        // @ts-ignore
        this.displayContent = content.style.display; //todo - fix
        content.style.display = "none";
        this.titleElement = new H4Element({
            text: title,
            padding: "0px 10px",
            order: "1",
            margin: "0px",
        });
        this.appendChild(this.titleElement);
        this.className = "tab-title";
        this.style.display = "grid";
        this.style.gridTemplateColumns = "auto max-content auto";
        this.style.gridTemplateRows = "1fr";
        this.style.alignItems = "center";
        this.closeable = closeable;
    }
    get closeable() {
        return this.__closeable;
    }
    set closeable(value) {
        this.__closeable = value;
        if (value && !this.closeIcon) {
            // @ts-ignore
            this.closeIcon = new IconElement({ icon: "fa-thin fa-x", order: "2" });
            this.closeIcon.onclick = () => {
                console.log("close");
                this.remove();
            };
            this.appendChild(this.closeIcon);
        }
        else {
            this.closeIcon?.remove();
            this.closeIcon = null;
        }
    }
    select() {
        if (!this.classList.contains("tab-title-active"))
            this.addClasName("tab-title-active");
        this._content.style.display = this.displayContent;
    }
    unselect() {
        if (this.classList.contains("tab-title-active"))
            this.classList.remove("tab-title-active");
        this._content.style.display = "none";
    }
    get content() {
        return this._content;
    }
}
// @ts-ignore
customElements.define("tab-item", TabItem);
